cmake/introspection.cmake

```diff
@@ -0,0 +1,255 @@
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+include(CheckCXXSourceCompiles)
+include(CheckCXXSymbolExists)
+include(CheckIncludeFileCXX)
+include(TestBigEndian)
+
+test_big_endian(WORDS_BIGENDIAN)
+
+# The following HAVE_{HEADER}_H variables go to the bitcoin-config.h header.
+CHECK_INCLUDE_FILE_CXX(byteswap.h HAVE_BYTESWAP_H)
+CHECK_INCLUDE_FILE_CXX(endian.h HAVE_ENDIAN_H)
+CHECK_INCLUDE_FILE_CXX(sys/endian.h HAVE_SYS_ENDIAN_H)
+CHECK_INCLUDE_FILE_CXX(sys/prctl.h HAVE_SYS_PRCTL_H)
+CHECK_INCLUDE_FILE_CXX(sys/resources.h HAVE_SYS_RESOURCES_H)
+CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+CHECK_INCLUDE_FILE_CXX(sys/sysctl.h HAVE_SYS_SYSCTL_H)
```

Comment:
We check for `sys/sysctl.h` here, and set `HAVE_SYS_SYSCTL_H`, but then never use `HAVE_SYS_SYSCTL_H`, while including `sys/sysctl.h` unconditionally inside CMake, and under various different identifiers in our src, i.e `HAVE_SYSCTL`, `HAVE_SYSCTL_ARND`, `MAC_OSX && ENABLE_ARM_SHANI`.

Similar for `sys/select.h`. We do a header check, but then never use `HAVE_SYS_SELECT_H` and just include `sys/select.h` in a `#ifndef WIN32` in compat.

I think now would be the right time to introduce some consistency. i.e we could drop the header checks here, and assume that if we can use `sysctl()`/`sysctlbyname()`, then we can include `sys/sysctl.h`, and use the same define for both, or, we could actually use the output of the header checks to include the headers. I don't think we should port checks that create unused output, and then go and do a number of different things in the source anyways. 

Doing this for a "similar bitcoin-config.h" is fine (either initially/during first review/porting), but no point bringing this all in at merge-time if it's unused, or not without a plan for how we are going to consolidate/remove it all.

Reply:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436835797).